SCD2 support

jorritsandbrink opened the pull request:
<!--
Thank you for submitting a pull request! Please provide a brief description of your changes below.
-->
### Description
This PR adds basic support for SCD2. Scope is defined [here](https://github.com/dlt-hub/dlt/issues/828#issuecomment-2016791397).

- Treats `scd2` as a special case of the `merge` write disposition.
- Support configurable names for the validity columns ("valid from" / "valid to").
- Validity column names default to `_dlt_valid_from` and `_dlt_valid_to` if the user does not specify names.
- Uses timestamps (not dates) to support sub-daily changes.
- Uses high timestamp (`9999-12-31T00:00:00+00:00`) in "valid to" column to indicate active records.
- Includes validity columns only in the root table. Validity dates for records in child tables can be linked through the root id.
- Does **not** use "standard" [configuration](https://dlthub.com/docs/general-usage/credentials/configuration) because it doesn't support resource-scoped configuration (correct me if I'm wrong). We want resource-scoped configuration because a user might want to load one table with the `scd2` strategy and another one with "regular" merge behavior, or use different validity column names for different tables.
- ~~Introduces `merge_config` argument on the `resource` decorator~~ Extends `write_disposition` argument such that it also accepts configuration dictionaries to let the user specify the merge strategy and validity column names..
- ~~Adds `merge_config` key to `TResourceHints`, but not to `TTableSchema`.~~
- Uses x-hints to store user configuration in the schema:
  - `x-merge-strategy` at the table level
  - `x-valid-from` and `x-valid-to` at the column level

<!--
Please link any related issues. This helps us keep the PR focused and merge it faster.
-->
### Related Issues

- Closes #828

sultaniman commented on tests/load/pipeline/test_scd2.py:
```diff
@@ -0,0 +1,370 @@
+
+
+ACTIVE_TS = datetime.fromisoformat(HIGH_TS.isoformat()).replace(tzinfo=None)
+h = DataItemNormalizer.get_row_hash
```
maybe alias this to `get_row_hash`?

jorritsandbrink replied:
Changed in https://github.com/dlt-hub/dlt/pull/1168/commits/7726d98a5a412da45b6a6a1f9bc3ffb736b781cd

sultaniman commented on dlt/common/schema/typing.py:
```diff
@@ -155,6 +155,14 @@ class NormalizerInfo(TypedDict, total=True):
 
+class TMergeConfig(TypedDict, total=False):
+    strategy: Optional[TLoaderMergeStrategy]
+    validity_column_names: Optional[List[str]]
```
Should it allow duplicates column names?

jorritsandbrink replied:
It shouldn't, didn't think of that. Added validity column name checking in https://github.com/dlt-hub/dlt/pull/1168/commits/30bb2e0e9fff508a76c9cea56aed997c849cf377. An exception is raised if a configured validity column name appears in the data.

sultaniman commented on dlt/destinations/sql_jobs.py:
```diff
@@ -478,3 +491,61 @@ def gen_merge_sql(
+
+        # insert new active records in root table
+        columns = map(escape_id, list(root_table["columns"].keys()))
+        col_str = ", ".join([c for c in columns if c not in (from_, to)])
```
afaik there are requests to preserve column names and in the future we might want to support it, should we put column names inside quotes?

jorritsandbrink replied:
I think we need to escape column names. Without escaping, for example, the column name `from` (a logical candidate name for the "valid from" column) would cause an error because it's a SQL keyword.

rudolfix replied:
there's another PR that will handle that @sultaniman there will be proper escaping and normalization of column names

sultaniman commented on tests/load/pipeline/test_scd2.py:
```diff
@@ -0,0 +1,370 @@
+)
+
+
+ACTIVE_TS = datetime.fromisoformat(HIGH_TS.isoformat()).replace(tzinfo=None)
```
Can't we just use `HIGHT_TS.replace(tzinfo=None)`?

jorritsandbrink replied:
`HIGH_TS` is a `pendulum` `DateTime` because the `datetime` module is a banned import. However, when fetching data from the destination tables to assert its content, timestamps are returned as `datetime` objects. Hence the conversion from `DateTime` to `datetime.`

sultaniman commented on dlt/destinations/sql_jobs.py:
```diff
@@ -478,3 +491,61 @@ def gen_merge_sql(
+            WHERE NOT EXISTS (
+                SELECT s._dlt_id FROM {staging_root_table_name} AS s
+                WHERE {root_table_name}._dlt_id = s._dlt_id
+            ) AND {to} = '{active_record_ts}';
```
with this condition aren't we making a useless update?

sh-rp replied:
yes this looks totally fine for me too

sh-rp commented on tests/load/pipeline/test_scd2.py:
```diff
@@ -0,0 +1,399 @@
+    ids=lambda x: x.name,
+)
+@pytest.mark.parametrize("simple", [True, False])
+def test_child_table(destination_config: DestinationTestConfiguration, simple: bool) -> None:
```
i think we should test the child table functionality on all destinations, so either enable all dests here or add a child table to the basic test. the reason is, that we should run all sql code that we have against all destinations, just to make sure it works everywhere.

jorritsandbrink replied:
Done in https://github.com/dlt-hub/dlt/pull/1168/commits/11748a63ffee50d2f01eb80a6138444a56743980

sh-rp reviewed (changes requested):
Very nice work, thanks! Only a few small changes and we still need the docs. Also I want to discuss two points with @rudolfix before we can merge.

jorritsandbrink commented:
> @rudolfix two questions on this PR:
> 
> * Do you agree with the interface? We now have additional fields on the resource to configure this as opposed to a settings on the destination how we do it with the replace strategy. I think this makes sense, but also looks a bit weird provided the replace strategy is set in a different way.
> * Right now the "cutoff date" is the created date of the load package. I am thinking we might want to make this overridable with a setting for certain cases, but maybe it is good like this now and we can add this feature when people request it.

Note that this is why I chose a different interface than the one used to configure replace strategy:
![image](https://github.com/dlt-hub/dlt/assets/47451109/c014792e-5809-4807-afee-48f605bbe0cf)

rudolfix commented on dlt/destinations/sql_jobs.py:
```diff
@@ -478,3 +492,60 @@ def gen_merge_sql(
+        from_, to = list(map(escape_id, get_validity_column_names(root_table)))
+
+        # define values for validity columns
+        boundary_ts = current_load_package()["state"]["created_at"]
```
looks good for now!

rudolfix commented on dlt/common/schema/typing.py:
```diff
@@ -65,6 +64,7 @@
 ]
 """Known hints of a column used to declare hint regexes."""
 TWriteDisposition = Literal["skip", "append", "replace", "merge"]
+TLoaderMergeStrategy = Literal["scd2"]
```
we should have a default strategy which is "delete-insert". and we'll add one more `merge` to support MERGE sql statements #1129

jorritsandbrink replied:
Added in https://github.com/dlt-hub/dlt/pull/1168/commits/396ec59393f691573d68585befd99abcbb4fc5c8

rudolfix commented on dlt/extract/decorators.py:
```diff
@@ -297,6 +297,7 @@ def resource(
     columns: TTableHintTemplate[TAnySchemaColumns] = None,
     primary_key: TTableHintTemplate[TColumnNames] = None,
     merge_key: TTableHintTemplate[TColumnNames] = None,
+    merge_config: TMergeConfig = None,
```
@jorritsandbrink @sh-rp maybe instead of doing this, we could extend `write_disposition` to
``
write_disposition: TTableHintTemplate[Union[TWriteDisposition, TMergeConfig, TReplaceConfig]] = None
```
exactly what we do with `schema_contract` below? then we can support short hand strings and full definitions. and still use the same parameter
in that case I'd rename TMergeConfig to TMergeDispositionConfig

sh-rp replied:
and we also handle the replace strategies in there? (not in the pr but later)? i think it's a good idea

jorritsandbrink replied:
Extended `write_disposition` in https://github.com/dlt-hub/dlt/pull/1168/commits/8f4d4cee6e59e37962963670bf2d23a1d1454e99 and https://github.com/dlt-hub/dlt/pull/1168/commits/c8d84d88a8fc8e221cb3b60c8930be77be67009f

rudolfix reviewed (changes requested):
this is really good! @jorritsandbrink congrats on plugging the work into our normalizer and hints perfectly.

I think we can do the user interface better by extending `write_disposition` instead of adding a new argument. @sh-rp exactly what we did for data contracts with shorthand and full notation - see my review

also: I trust you that SQL transformation is good and properly tested :)

jorritsandbrink commented:
@rudolfix I adjusted the user interface according to your suggestion and introduced a default `delete-insert` merge strategy. WDYT?

I'll add docs after we've settled on the user interface.

rudolfix reviewed (changes requested):
this is good! I have two optimizations that IMO are OK to add.
thx for adding apply_hints tests!

rudolfix commented on dlt/destinations/sql_jobs.py:
```diff
@@ -486,3 +499,60 @@ def gen_merge_sql(
+            UPDATE {root_table_name} SET {to} = '{boundary_ts}'
+            WHERE NOT EXISTS (
+                SELECT s._dlt_id FROM {staging_root_table_name} AS s
+                WHERE {root_table_name}._dlt_id = s._dlt_id
```
let's add x-row-hash hint that you can attach to `_dlt_id` if it not present in the schema (same thing we do with `root_key`). why:
1. no dependency on particular name when generating sql but on hints
2. users can create their own hash during extraction and ie. just take into account 2-3 columns not all of them. also our default hash (via json dump) will be quite slow

jorritsandbrink replied:
We could handle three user input cases:
1. `merge_key` = `None` (no merge key) ➜ assign `x-row-hash` to `_dlt_id` and generate row hash including all resource columns (the ones not generated by `dlt`)
2. `merge_key` = `col_x` (simple merge key) ➜ assign `x-row-hash` to `col_x` and use random `_dlt_id`—this can be used as "Bring Your Own Hash / Surrogate Key"
3. `merge_key` = [`col_x`, `col_y`, `col_z`] (composite merge key) ➜ assign `x-row-hash` to `_dlt_id` and generate row hash including only `col_x`, `col_y`,  and `col_z`

I suppose we would need to restrict `merge_key` changes. If we don't, we would identify unchanged records as being updated because the way their hash was calculated changed.

jorritsandbrink replied:
Added `row_hash_column_name` configuration option in https://github.com/dlt-hub/dlt/pull/1168/commits/4124d61574d87c42230c33913d1373603af527f5. Translates to `x-row-hash` hint behind the scenes. With this, users can bring their own hash. As discussed on Slack, let's move other extensions involving `merge_key` to follow up PRs.

rudolfix reviewed (changes requested):
@jorritsandbrink what should we do with `merge_key` here? should we warn/raise when it is specified or use it for something ie. to apply the scd2 to a subset of data defined by a merge_key?
in the second case it is still possible to do incremental ie. day by updated

jorritsandbrink commented on dlt/common/normalizers/json/relational.py:
```diff
@@ -296,10 +319,18 @@ def normalize_data_item(
         row = cast(TDataItemRowRoot, item)
         # identify load id if loaded data must be processed after loading incrementally
         row["_dlt_load_id"] = load_id
+        # determine if row hash should be used as dlt id
```
Solved with caching as you suggested on Slack: https://github.com/dlt-hub/dlt/pull/1168/commits/6b243782f3f53dc8fb6e8632795073c4dc110dcb

jorritsandbrink commented on dlt/extract/hints.py:
```diff
@@ -391,17 +399,56 @@ def _merge_key(hint: TColumnProp, keys: TColumnNames, partial: TPartialTableSche
+    def _create_table_schema(resource_hints: TResourceHints, resource_name: str) -> TTableSchema:
+        """Creates table schema from resource hints and resource name."""
+
+        dict_ = cast(Dict[str, Any], deepcopy(resource_hints))
```
`deepcopy` was indeed obsolete—removed it in https://github.com/dlt-hub/dlt/pull/1168/commits/1f399bcf5c306c036e643ce82d291755974737b3

rudolfix reviewed (dismissed):
LGTM!

oh `dremio` does not like ZULU timestamps in sql merge job... do you generate them for SCD2? I think Z timestamps are not really part of ISO

we have a dremio container in tests if you want to fix it
